tests/accelerators/test_ipu.py

```diff
@@ -0,0 +1,504 @@
+    assert isinstance(trainer.accelerator, IPUAccelerator)
+
+
+@RunIf(ipu=True)
+@pytest.mark.parametrize('ipus', [1, 4])
+def test_all_stages(tmpdir, ipus):
+    model = IPUModel()
+    trainer = Trainer(fast_dev_run=True, ipus=ipus)
+    trainer.fit(model)
+    trainer.validate(model)
+    trainer.test(model)
+    trainer.predict(model, model.val_dataloader())
+
+
+@RunIf(ipu=True)
+@pytest.mark.parametrize('ipus', [1, 4])
+def test_inference_only(tmpdir, ipus):
+    model = IPUModel()
+
+    trainer = Trainer(fast_dev_run=True, ipus=ipus)
+    trainer.validate(model)
+    trainer.test(model)
+    trainer.predict(model, model.val_dataloader())
+
+
+@RunIf(ipu=True)
+def test_optimization(tmpdir):
+    seed_everything(42)
+
+    dm = ClassifDataModule(length=1024)
+    model = IPUClassificationModel()
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=1,
+        ipus=2,
+    )
+
+    # fit model
+    trainer.fit(model, dm)
+    assert trainer.state.finished, f"Training failed with {trainer.state}"
+    assert dm.trainer is not None
+
+    # validate
+    result = trainer.validate(datamodule=dm)
+    assert dm.trainer is not None
+    assert result[0]['val_acc'] > 0.7
+
+    # test
+    result = trainer.test(model, datamodule=dm)
+    assert dm.trainer is not None
+    test_result = result[0]['test_acc']
+    assert test_result > 0.6
+
+    # test saved model
+    model_path = os.path.join(tmpdir, 'model.pt')
+    trainer.save_checkpoint(model_path)
+
+    model = IPUClassificationModel.load_from_checkpoint(model_path)
+
+    trainer = Trainer(default_root_dir=tmpdir, ipus=2)
+
+    result = trainer.test(model, datamodule=dm)
+    saved_result = result[0]['test_acc']
+    assert saved_result == test_result
+
+
+@RunIf(ipu=True)
+def test_mixed_precision(tmpdir):
+
+    class TestCallback(Callback):
+
+        def setup(self, trainer: Trainer, pl_module: LightningModule, stage: Optional[str] = None) -> None:
+            assert isinstance(trainer.accelerator.precision_plugin, IPUPrecisionPlugin)
+            assert trainer.accelerator.precision_plugin.precision == 16
+            assert trainer.accelerator.model.precision == 16
+            raise SystemExit
+
+    model = IPUModel()
+    trainer = Trainer(fast_dev_run=True, ipus=1, precision=16, callbacks=TestCallback())
+    with pytest.raises(SystemExit):
+        trainer.fit(model)
+
+
+@RunIf(ipu=True)
+def test_pure_half_precision(tmpdir):
+
+    class TestCallback(Callback):
+
+        def on_train_start(self, trainer: Trainer, pl_module: LightningModule) -> None:
+            assert isinstance(trainer.accelerator.training_type_plugin, IPUPlugin)
+            assert isinstance(trainer.accelerator.precision_plugin, IPUPrecisionPlugin)
+            assert trainer.accelerator.precision_plugin.precision == 16
+            assert trainer.accelerator.model.precision == 16
+            assert trainer.accelerator.training_type_plugin.convert_model_to_half
```

Comment:
we should be able to run these asserts (or at least most of them) right after the trainer init right?

Reply:
Let me try!